### PR TITLE
kernel-build.eclass: Fix separate private and public module signing keys

### PR DESCRIPTION
**Pull request by vimproved**
The kernel expects CONFIG_MODULE_SIG_KEY to be either a pkcs11 URI containing refences to both a private and public key, or a path to a PEM file containing both the private and public keys. However, currently the kernel build will fail if MODULES_SIGNING_KEY is set to a PEM file containing only the private key. This commit adds a step in kernel-build_merge_configs that concatenates MODULES_SIGNING_KEY and MODULES_SIGNING_CERT into ${T}/kernel_key.pem if both files exist and are not the same path. It then sets MODULES_SIGNING_KEY to ${T}/kernel_key.pem. This should fix building with separate private and public module signing keys. (ie if the user created their signing keys as described in https://wiki.gentoo.org/wiki/Signed_kernel_module_support)

### CODE REVIEW

File: eclass/kernel-build.eclass [hunk 1]
```diff
@@ -89,6 +90,14 @@ IUSE="+strip"
...
 # Default if unset: certs/signing_key.pem
 
+# @ECLASS_VARIABLE: MODULES_SIGN_CERT
+# @USER_VARIABLE
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Used with USE=modules-sign.  Can be set to the path of the public
+# key in PEM format to use. Must be specified if MODULES_SIGN_KEY
+# is set to a path to a file that does only contains the private key.
+
```
Comment (AndrewAmmerlaan) [inline eclass/kernel-build.eclass:100]:
```suggestion
# is set to a path of a file that only contains the private key.

```

**Review by AndrewAmmerlaan — APPROVED** [review]
Conceptually LGTM. Haven't actually tested yet because I'm currently traveling. I wonder if we can somehow reliably detect when a PEM file only contains the key and die with a nice error message if this is the case. This would be useful because otherwise you'll only find out that your configuration doesn't work *after* you've compiled the kernel which is a bit of a waste of time.

**Comment by AndrewAmmerlaan** [comment]
Now I've also tested this locally, works well for me :+1: 

@vimproved since this is a patch for an eclass it must also be sent to the mailing list for further review. I just sent it out, and put you in the CC.

**Comment by vimproved** [comment]
> Now I've also tested this locally, works well for me 👍
> 
> @vimproved since this is a patch for an eclass it must also be sent to the mailing list for further review. I just sent it out, and put you in the CC.

Ah, thanks. Forgot about that.

**Comment by AndrewAmmerlaan** [comment]
@vimproved did you see @mgorny 's comments via e-mail? Could you address his feedback, then I'll merge this afterwards.

File: eclass/kernel-build.eclass [hunk 2]
```diff
@@ -394,6 +403,12 @@ kernel-build_merge_configs() {
 				CONFIG_MODULE_SIG_FORCE=y
 				CONFIG_MODULE_SIG_${MODULES_SIGN_HASH^^}=y
 			EOF
+			if [[ -e ${MODULES_SIGN_KEY} ]] && [[ -e ${MODULES_SIGN_CERT} ]] \
+				&& [[ ${MODULES_SIGN_KEY} != ${MODULES_SIGN_CERT} ]] \
+				&& [[ ${MODULES_SIGN_KEY} != pkcs11:* ]]; then
```
Comment (AndrewAmmerlaan) [inline eclass/kernel-build.eclass:408]:
```suggestion
			if [[ -e ${MODULES_SIGN_KEY} && -e ${MODULES_SIGN_CERT} &&
				${MODULES_SIGN_KEY} != ${MODULES_SIGN_CERT} &&
				${MODULES_SIGN_KEY} != pkcs11:* ]]; then
```

File: eclass/kernel-build.eclass [hunk 3]
```diff
@@ -394,6 +403,12 @@ kernel-build_merge_configs() {
 				CONFIG_MODULE_SIG_FORCE=y
 				CONFIG_MODULE_SIG_${MODULES_SIGN_HASH^^}=y
 			EOF
+			if [[ -e ${MODULES_SIGN_KEY} && -e ${MODULES_SIGN_CERT} &&
+				${MODULES_SIGN_KEY} != ${MODULES_SIGN_CERT} &&
+				${MODULES_SIGN_KEY} != pkcs11:* ]]; then
```
Comment (mgorny) [inline eclass/kernel-build.eclass:408]:
```suggestion
				${MODULES_SIGN_KEY} != ${MODULES_SIGN_CERT} &&
				${MODULES_SIGN_KEY} != pkcs11:* ]]
			then
```

to keep the visual separation.

**Review by mgorny — APPROVED** [review]
Thanks!

@AndrewAmmerlaan, can you confirm before I merge?

**Comment by AndrewAmmerlaan** [comment]
> Thanks!
> 
> @AndrewAmmerlaan, can you confirm before I merge?

Still works great :+1:

**Pull request closed without merging**